File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int len = Integer.parseInt(br.readLine());
        Stack<String> stack = new Stack<>();
        while(len-- > 0) {
            String[] arr = br.readLine().split("");
            for(String a : arr) {
                if(a.equals("(")) {
                    stack.push(a);
                } else if(a.equals(")")) {
                    if(!stack.isEmpty() && stack.peek().equals("(")) {
                        stack.pop();
                    } else {
                        stack.push(a);
                    }
                }
            }
            if (stack.isEmpty()) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
            stack.clear();
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline


for _ in range(int(input())):
    arr = []
    td = input().strip()

    for t in td:
        if not len(arr):
            arr.append(t)
            continue

        if t == ')' and arr[-1] == '(':
            arr.pop()
        else:
            arr.append(t)
    print('YES' if not len(arr) else 'NO')

